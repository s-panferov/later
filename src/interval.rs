use chrono::{DateTime, Duration, Utc};
use std::fmt::Debug;

#[derive(Debug, Clone)]
pub struct Interval {
	pub from: DateTime<Utc>,
	pub to: Option<DateTime<Utc>>,
}

impl Interval {
	pub fn from_now() -> Self {
		Interval {
			from: Utc::now(),
			to: None,
		}
	}

	pub fn from(from: DateTime<Utc>) -> Self {
		Interval { from, to: None }
	}
}

/// Something that resolves to the range of time
pub trait AsIntervals: Debug {
	/// Get the estimate of the interval
	fn duration_hint(&self) -> Duration;

	/// Get an interator to resolve intervals
	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>>;
}
