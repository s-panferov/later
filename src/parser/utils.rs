use nom::{
	branch::alt,
	bytes::complete::tag,
	character::complete::{space0, space1},
	combinator::opt,
	sequence::tuple,
};

use super::error::ParseResult;

pub fn parse_chain<T, F>(input: &str, parser: F) -> ParseResult<Vec<T>>
where
	F: Fn(&str) -> ParseResult<T>,
{
	let mut res = vec![];
	let sep = opt(alt((
		tuple((space0, tag(","), space0)),
		tuple((space1, tag("and"), space1)),
	)));

	let (mut input, value) = parser(input)?;
	res.push(value);

	loop {
		let (i, s) = sep(input)?;
		match s {
			Some(_) => {
				let (i, value) = parser(i)?;
				res.push(value);
				input = i
			}
			None => break,
		}
	}

	Ok((input, res))
}
