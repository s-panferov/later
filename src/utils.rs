use chrono::{DateTime, Datelike, NaiveDate, Timelike, Utc};

pub enum Of {
	Day,
	Month,
	Year,
}

pub fn end_of(of: Of, date: &DateTime<Utc>) -> DateTime<Utc> {
	match of {
		Of::Year => end_of(Of::Month, &date.with_month(12).unwrap()),
		Of::Month => end_of(
			Of::Day,
			&date
				.with_day(days_in_month(date.year(), date.month()))
				.unwrap(),
		),
		Of::Day => date
			.with_hour(23)
			.unwrap()
			.with_minute(59)
			.unwrap()
			.with_second(59)
			.unwrap()
			.with_nanosecond(999999999)
			.unwrap(),
	}
}

pub fn start_of(of: Of, date: &DateTime<Utc>) -> DateTime<Utc> {
	match of {
		Of::Year => start_of(Of::Month, &date.with_month(1).unwrap()),
		Of::Month => start_of(Of::Day, &date.with_day(1).unwrap()),
		Of::Day => date
			.with_hour(0)
			.unwrap()
			.with_minute(0)
			.unwrap()
			.with_second(0)
			.unwrap()
			.with_nanosecond(0)
			.unwrap(),
	}
}

pub fn days_in_month(year: i32, m: u32) -> u32 {
	if m == 12 {
		NaiveDate::from_ymd(year + 1, 1, 1)
	} else {
		NaiveDate::from_ymd(year, m + 1, 1)
	}
	.signed_duration_since(NaiveDate::from_ymd(year, m, 1))
	.num_days() as u32
}
