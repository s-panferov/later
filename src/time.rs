use chrono::Timelike;
use chrono::{Duration, NaiveTime};

use crate::interval::{Interval, Timeline};
use crate::utils::{end_of, start_of, Of};

impl Timeline for NaiveTime {
	fn duration_hint(&self) -> Duration {
		Duration::days(1)
	}

	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>> {
		Box::new(TimeIterator {
			time: self.clone(),
			interval,
			completed: false,
		})
	}
}

struct TimeIterator {
	time: NaiveTime,
	interval: Interval,
	completed: bool,
}

impl Iterator for TimeIterator {
	type Item = Interval;

	fn next(&mut self) -> Option<Interval> {
		if self.completed {
			return None;
		}

		let mut from = self.interval.from.clone();
		let from_time = from.time();

		if from_time > self.time {
			// Jump to the next day
			from = start_of(
				Of::Day,
				&(end_of(Of::Day, &from) + Duration::seconds(1)),
			);
		}

		let datetime = from
			.with_hour(self.time.hour())
			.and_then(|time| time.with_minute(self.time.minute()))
      .and_then(|time| time.with_second(self.time.second()))
			.unwrap();

		if self.interval.to.is_some()
			&& datetime > *self.interval.to.as_ref().unwrap()
		{
			self.completed = true;
			return None;
		}

		self.interval.from = datetime + Duration::seconds(1);

		Some(Interval {
			from: datetime,
			to: Some(datetime),
		})
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::interval::{Interval, Timeline};
	use insta::assert_debug_snapshot_matches;

	#[test]
	fn every_10_am() {
		let am_10 = NaiveTime::from_hms(10, 0, 0);

		assert_debug_snapshot_matches!(
			"every_10_am",
			am_10
				.iter_within(Interval::from(
					"2019-10-31T12:00:00Z".parse().unwrap(),
				))
				.take(10)
				.collect::<Vec<_>>()
		);
	}
}
