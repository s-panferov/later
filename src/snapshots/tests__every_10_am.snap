---
created: "2019-09-05T05:08:22.943165Z"
creator: insta@0.10.1
source: src/time.rs
expression: "am_10.iter_within(Interval::from(\"2019-10-31T12:00:00Z\".parse().unwrap())).take(10).collect::<Vec<_>>()"
---
[
    Interval {
        from: 2019-11-01T10:00:00Z,
        to: Some(
            2019-11-01T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-02T10:00:00Z,
        to: Some(
            2019-11-02T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-03T10:00:00Z,
        to: Some(
            2019-11-03T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-04T10:00:00Z,
        to: Some(
            2019-11-04T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-05T10:00:00Z,
        to: Some(
            2019-11-05T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-06T10:00:00Z,
        to: Some(
            2019-11-06T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-07T10:00:00Z,
        to: Some(
            2019-11-07T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-08T10:00:00Z,
        to: Some(
            2019-11-08T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-09T10:00:00Z,
        to: Some(
            2019-11-09T10:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-10T10:00:00Z,
        to: Some(
            2019-11-10T10:00:00Z,
        ),
    },
]
